File: homework01-go/vigenere/vigenere.go
package vigenere

import "unicode"
import "unicode/utf8"

func EncryptVigenere(plaintext string, keyword string) string {
	var ciphertext string
	keywordLength := utf8.RuneCountInString(keyword)

	for index, character := range plaintext {
		if unicode.IsLower(character) {
			chrCode := int('a')
			shift := int(unicode.ToLower(rune(keyword[index%keywordLength]))) - chrCode
			ciphertext += string(rune(chrCode + (int(character)-chrCode+shift)%26))
		} else if unicode.IsUpper(character) {
			chrCode := int('A')
			shift := int(unicode.ToUpper(rune(keyword[index%keywordLength]))) - chrCode
			ciphertext += string(rune(chrCode + (int(character)-chrCode+shift)%26))
		} else {
			ciphertext += string(character)
		}
	}

	return ciphertext
}

func DecryptVigenere(ciphertext string, keyword string) string {
	var plaintext string
	keywordLength := utf8.RuneCountInString(keyword)

	for index, character := range ciphertext {
		if unicode.IsLower(character) {
			chrCode := int('a')
			shift := int(unicode.ToLower(rune(keyword[index%keywordLength]))) - chrCode
			newLetter := (int(character) - chrCode - shift) % 26
			if newLetter < 0 {
				newLetter += 26
			}
			plaintext += string(rune(chrCode + newLetter))
		} else if unicode.IsUpper(character) {
			chrCode := int('A')
			shift := int(unicode.ToUpper(rune(keyword[index%keywordLength]))) - chrCode
			newLetter := (int(character) - chrCode - shift) % 26
			if newLetter < 0 {
				newLetter += 26
			}
			plaintext += string(rune(chrCode + newLetter))
		} else {
			plaintext += string(character)
		}
	}

	return plaintext
}

File: homework01/vigenere.py
def encrypt_vigenere(plaintext, keyword):
    """
    Encrypts plaintext using a Vigenere cipher.

    >>> encrypt_vigenere("PYTHON", "A")
    'PYTHON'
    >>> encrypt_vigenere("python", "a")
    'python'
    >>> encrypt_vigenere("ATTACKATDAWN", "LEMON")
    'LXFOPVEFRNHR'
    """
    ciphertext = ''
    keyword_length = len(keyword)

    for index, character in enumerate(plaintext):
        if character.islower():
            chr_code = ord('a')
            shift = ord(keyword[index % keyword_length].lower()) - chr_code
            ciphertext += chr(chr_code + (ord(character) - chr_code + shift) % 26)
        elif character.isupper():
            chr_code = ord('A')
            shift = ord(keyword[index % keyword_length].upper()) - chr_code
            ciphertext += chr(chr_code + (ord(character) - chr_code + shift) % 26)
        else:
            ciphertext += character

    return ciphertext


def decrypt_vigenere(ciphertext, keyword):
    """
    Decrypts a ciphertext using a Vigenere cipher.

    >>> decrypt_vigenere("PYTHON", "A")
    'PYTHON'
    >>> decrypt_vigenere("python", "a")
    'python'
    >>> decrypt_vigenere("LXFOPVEFRNHR", "LEMON")
    'ATTACKATDAWN'
    """
    plaintext = ''
    keyword_length = len(keyword)

    for index, character in enumerate(ciphertext):
        if character.islower():
            chr_code = ord('a')
            shift = ord(keyword[index % keyword_length].lower()) - chr_code
            plaintext += chr(chr_code + (ord(character) - chr_code - shift) % 26)
        elif character.isupper():
            chr_code = ord('A')
            shift = ord(keyword[index % keyword_length].upper()) - chr_code
            plaintext += chr(chr_code + (ord(character) - chr_code - shift) % 26)
        else:
            plaintext += character
    return plaintext
